src/viam/sdk/components/encoder/encoder.hpp

```diff
@@ -0,0 +1,114 @@
+/// @file components/encoder/encoder.hpp
+///
+/// @brief Defines a `Encoder` component.
+#pragma once
+
+#include <string>
+
+#include <viam/api/component/encoder/v1/encoder.pb.h>
+
+#include <viam/sdk/common/proto_type.hpp>
+#include <viam/sdk/common/utils.hpp>
+#include <viam/sdk/config/resource.hpp>
+#include <viam/sdk/registry/registry.hpp>
+#include <viam/sdk/resource/resource_manager.hpp>
+
+namespace viam {
+namespace sdk {
+
+/// @defgroup Encoder Classes related to the `Encoder` component.
+
+/// @class EncoderSubtype
+/// @brief Defines a `ResourceSubtype` for the `Encoder` component.
+/// @ingroup Encoder
+class EncoderSubtype : public ResourceSubtype {
+   public:
+    std::shared_ptr<ResourceServerBase> create_resource_server(
+        std::shared_ptr<ResourceManager> manager) override;
+    std::shared_ptr<ResourceBase> create_rpc_client(std::string name,
+                                                    std::shared_ptr<grpc::Channel> chan) override;
+    EncoderSubtype(const google::protobuf::ServiceDescriptor* service_descriptor)
+        : ResourceSubtype(service_descriptor){};
+};
+
+/// @class Encoder encoder.hpp "components/encoder/encoder.hpp"
+/// @brief An encoder is a device that is hooked up to motors to report a position
+/// @ingroup Encoder
+///
+/// This acts as an abstract base class to be inherited from by any drivers representing
+/// specific encoder implementations. This class cannot be used on its own.
+class Encoder : public ComponentBase {
+   public:
+    /// @enum position_type
+    enum position_type {
+        // Unspecified position type
+        UNSPECIFIED,
+        // Provided by incremental encoders
+        TICKS_COUNT,
+        // Provided by absolute encoders
+        ANGLE_DEGREES
```

Comment:
Just whining here, but I'd love to switch the convention for constants (including enumerator values) in this project away from all-caps `CONSTANTS`. The capitalization of constants is from the old C days when const-ness wasn't a thing and the preprocessor was a major player. https://accu.org/journals/overload/22/121/wakely_1923/. Personally I like the Google style guide convention of prefixing constants with `k`, so `k_ticks_count` or `k_angles_degrees` because I do find it useful to be able to tell at a glance that a given symbol is a constant. But the all-caps is definitely distracting.

Reply:
Thank you for exposing me to this bit of history.

[Cpp reference](https://en.cppreference.com/w/cpp/language/enum) says `ticks_count` and `angle_degrees`.

[Google style guide](https://google.github.io/styleguide/cppguide.html#Enumerator_Names) says `kTicksCount` and `kAngleDegrees`

I am leaning towards `position_type::angle_degrees` because `position_type::k_angle_degrees` seems like the `k` means something to the user and the risk of collision is not obvious to me. 

Am I wrong about this? Happy to switch to whatever.